# NanoConsensus: consensus prediction of RNA modifications from direct RNA nanopore sequencing data

*NanoConsensus* is an software to robustly identify putative RNA modified sites from direct RNA sequencing datasets. 

## Table of Contents  
- [General Description](#General-Description)
- [Detailed Description](#Detailed-Description)
- [Installation](#Installation)
- [Running the code](#Running-the-code)
- [Expected output](#Expected-output)
- [Citation](#Citation) 
- [Contact](#Contact) 


## General Description

NanoConsensus performs pairwise comparisons between two conditions (e.g. WT vs KO) using four different RNA modification detection softwares ([Epinano](https://github.com/enovoa/EpiNano), [Nanopolish](https://github.com/jts/nanopolish), [Tombo](https://github.com/nanoporetech/tombo) and [Nanocompore](https://github.com/tleonardi/nanocompore)) at per-transcript level. Then, it combines all results generated to produce a consensus prediction, which is more robust than those from individual softwares.  

![NanoConsensus_scheme](/NanoConsensus_scheme.png)

## Detailed Description

The first step of *NanoConsensus* analysis is to run the four different RNA modification detection algorithms in a pairwise manner - comparing a WT against an IVT sample through [NanoMod](https://biocorecrg.github.io/master_of_pores/nanomod.html) module from [Master of Pores](https://github.com/biocorecrg/master_of_pores). These algorithms will provide results at position and at transcript level. 

*NanoConsensus* extracts the data from a specific transcript and software, it then performs Z-Score normalization for each dataset provided. The candidate positions are then selected by the resulting Z-Score which must be higher than the provided threshold the default value of which is 5. This threshold is able to be changed at the users discretion. In the following step, candidate positions from all software are extended into 5-mers. Flexible overlapping is then performed, providing those regions supported by data from two or more softwares, which are then saved as putative modified sites. 

*NanoConsensus* uses the median of the rescaled Z-score, which is rescaled between 0 and 1, at a specific position from each of the four different softwares providing a nanconcensus score for that position. This is performed at all positions across the determined transript. 

In the last step, the *NanoConsensus scores* from all previously identified as putative modified sites are verified. This verification is performed by comparing the *Nanoconsensus score* to a threshold which is determined by the median of the *Nanoconsensus scores* across the entire transcript multiplied by an integer, with 5 as the default value. The decisive verified results are then reported whilst all unverified results are discarded.

## Installation 
If needed, install *Master of Pores* from its github repository:
```
git clone https://github.com/biocorecrg/master_of_pores_dev.git
```

Then, install *NanoConsensus* from its github repository:
```
git clone https://github.com/ADelgadoT/NanoConsensus.git
```

## Running the code 

### Required inputs 
To perform NanoConsensus analysis, user must provide the **following inputs**:
* Reference file (*.fa)
* Transcript name
* Start position
* End position

These inputs - generated by *NanoMod* - are also required for the analysis:
* **Epinano output**: both for WT and IVT samples
```
#Ref,pos,base,cov,q_mean,q_median,q_std,mis,ins,del
23S,2515,C,45497.0,5.36995,4.00000,3.97797,0.0822032221904741,0.18715519704595907,0.2058377475437941
23S,2516,A,45504.0,5.38207,4.00000,4.71619,0.17128164556962025,0.20497099156118143,0.07733386075949367
23S,2517,C,45529.0,6.92130,5.00000,5.04250,0.06165301236574491,0.1505633771881658,0.13540820136616222
23S,2518,A,45545.0,6.49821,5.00000,5.47485,0.10802503018992206,0.10855198155670216,0.2082775277198375
23S,2519,T,45557.0,6.51247,5.00000,4.81853,0.09386043857145993,0.14792457800118533,0.2033057488421099
23S,2520,C,45609.0,10.72107,9.00000,7.03936,0.03990440483237957,0.1465938740160933,0.020105680896314326
```

* **Nanopolish output**: both for WT and IVT samples
```
contig	position	reference_kmer	read_name	median	coverage
23S	2515	ACATC	1	79.03	271377
23S	2516	CATCC	1	87.0	118486
23S	2517	ATCCT	1	74.47	124448
23S	2518	TCCTG	1	73.21	65149
23S	2519	CCTGG	1	88.2	84488
23S	2520	CTGGA	1	107.84	125206
```

* **Tombo output**
```
"Ref_Position"	"Chr"	"Position"	"Tombo_SiteScore"	"Coverage_Sample"	"Coverage_IVT"	"Tombo_KmerScore"
"23S_2515"	"23S"	"2515"	"0.0000"	"37042"	"37884"	0
"23S_2516"	"23S"	"2516"	"0.0000"	"37039"	"37880"	5e-04
"23S_2517"	"23S"	"2517"	"0.0000"	"37053"	"37895"	0.0012
"23S_2518"	"23S"	"2518"	"0.0005"	"37050"	"37892"	0.0018
"23S_2519"	"23S"	"2519"	"0.0007"	"37064"	"37907"	0.0018
"23S_2520"	"23S"	"2520"	"0.0006"	"37141"	"37988"	0.0018
```

* **Nanocompore output**
```
pos	chr	genomicPos	ref_id	strand	ref_kmer	GMM_logit_pvalue	GMM_logit_pvalue_context_4	KS_dwell_pvalue	KS_dwell_pvalue_context_4	KS_intensity_pvalue	KS_intensity_pvalue_context_4	MW_dwell_pvalue	MW_dwell_pvalue_context_4	MW_intensity_pvalue	MW_intensity_pvalue_context_4	TT_dwell_pvalue	TT_dwell_pvalue_context_4	TT_intensity_pvalue	TT_intensity_pvalue_context_4	GMM_cov_type	GMM_n_clust	cluster_counts	Logit_LOR
2515	NA	NA	23S	NA	ACATC	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:2137/5121__JW3718_1:2284/5121	-0.06649521174331559
2516	NA	NA	23S	NA	CATCC	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:2946/4290__JW3718_1:3033/4347	-0.015898013921205018
2517	NA	NA	23S	NA	ATCCT	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:5699/1344__JW3718_1:5818/1374	0.001397467230704439
2518	NA	NA	23S	NA	TCCTG	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:159/6944__JW3718_1:150/7091	0.07883939042455489
2519	NA	NA	23S	NA	CCTGG	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:4179/2943__JW3718_1:4325/2949	-0.032296111509103595
2520	NA	NA	23S	NA	CTGGA	0.9999145725030434	0.9999997713740346	1.0	1.0	1.0	1.0	0.9998988895345491	0.9999961317288804	0.9997948160734869	0.9999943103243797	0.9998967130410295	0.999999547434852	0.9999171708631587	0.9999935368706631	full	2	Parent10_JW3718-90_1:3925/3177__JW3718_1:3981/3235	0.003922821293985011
```

### Usage

* Default command:
```
Rscript NanoConsensus.R -Epi_Sample /path/Epinano/WT/data -Epi_IVT /path/Epinano/IVT/data -NP_Sample /path/Nanopolish/WT/data -NP_IVT /path/Nanopolish/IVT/data -Tombo /path/Tombo/data -Nanocomp /path/Nanocompore/data -ini_pos start_position -fin_pos end_position -output output_name -fasta /path/reference/fasta -chr transcript_name 
```

* Use the following command to produce intermediate and final plots:
```
Rscript NanoConsensus.R -Epi_Sample /path/Epinano/WT/data -Epi_IVT /path/Epinano/IVT/data -NP_Sample /path/Nanopolish/WT/data -NP_IVT /path/Nanopolish/IVT/data -Tombo /path/Tombo/data -Nanocomp /path/Nanocompore/data -ini_pos start_position -fin_pos end_position -output output_name -fasta /path/reference/fasta -chr transcript_name -plot
```

* Use the following command to change the Z-Score threshold (default = *5*) to identify putative modified sites from individual software data:
```
Rscript NanoConsensus.R -Epi_Sample /path/Epinano/WT/data -Epi_IVT /path/Epinano/IVT/data -NP_Sample /path/Nanopolish/WT/data -NP_IVT /path/Nanopolish/IVT/data -Tombo /path/Tombo/data -Nanocomp /path/Nanocompore/data -ini_pos start_position -fin_pos end_position -output output_name -fasta /path/reference/fasta -chr transcript_name -plot --MZS_thr new_threshold
```

* Use the following command to change the NanoConsensus score threshold (default = median(NanoConsensus_Score) * *5*) to filter the final putative modified sites:
```
Rscript NanoConsensus.R -Epi_Sample /path/Epinano/WT/data -Epi_IVT /path/Epinano/IVT/data -NP_Sample /path/Nanopolish/WT/data -NP_IVT /path/Nanopolish/IVT/data -Tombo /path/Tombo/data -Nanocomp /path/Nanocompore/data -ini_pos start_position -fin_pos end_position -output output_name -fasta /path/reference/fasta -chr transcript_name -plot --NC_thr new_threshold
```

## Expected output 

**TO BE FILLED IN** 

## Citation

A pre-print is currently under preparation.

## Contact 

Please open an issue in the GitHub repo if you have any questions/doubts/suggestions about how to use this software. Thanks!

